tests/unit/resources/test_factory.py

```diff
@@ -686,6 +686,57 @@ def test_dangling_resource_inequality(self):
         self.assertNotEqual(q1, q2)
         self.assertNotEqual(q1, m)
 
+    def test_dangling_resource_loads_data(self):
+        # Given a loadable resource instance that contains a reference
+        # to another resource which has a resource data path, the
+        # referenced resource should be loaded with all of the data
+        # contained at that path. This allows loading references
+        # which would otherwise not be loadable (missing load method)
+        # and prevents extra load calls for others when we already
+        # have the data available.
+        self.defs = {
+            'Instance': {
+                'identifiers': [{'name': 'Id'}],
+                'has': {
+                    'NetworkInterface': {
+                        'resource': {
+                            'type': 'NetworkInterface',
+                            'identifiers': [
+                                {'target': 'Id', 'source': 'data', 'path': 'NetworkInterface.Id'}
+                            ],
+                            'path': 'NetworkInterface'
+                        }
+                    }
+                }
+            },
+            'NetworkInterface': {
+                'identifiers': [{'name': 'Id'}]
+            }
+        }
+        self.model = self.defs['Instance']
+
+        cls = self.load('test', 'Instance', self.model, self.defs, None)
+        instance = cls('instance-id')
+
+        # Set some data as if we had completed a load action.
+        def set_meta_data():
+            instance.meta.data = {
+                'NetworkInterface': {
+                    'Id': 'network-interface-id',
+                    'PublicIp': '127.0.0.1'
+                }
+        }
+        instance.load = mock.Mock(side_effect=set_meta_data)
+
+        # Now, get the reference and make sure it has its identifier
+        # and all its data set.
+        network_interface = instance.network_interface
+        self.assertEqual(network_interface.id, 'network-interface-id')
```

Comment:
Yeah that makes sense, but my concern is that I would prefer to have some sort of test that is closer to how a user would actually use this feature, i.e how you initially layed out the problem in the pull request description.


Maybe it's not worth modifying this test and it's more appropriate for a "functional" or integration test.  But at any rate, it seems like a good idea to have some type of test that verifies this behavior in terms of the user provided API.